src/box/lua/config/applier/roles.lua

```diff
@@ -180,8 +242,16 @@ local function post_apply(config)
         end
     end
 
-    last_loaded = loaded
-    last_roles_ordered = roles_ordered
+    roles_state.last_loaded = loaded
+    roles_state.last_roles_ordered = roles_ordered
+    roles_state.last_config = table.deepcopy(config)
+
+    -- box.status watcher values should be already set, because the watcher
+    -- has been registered during apply.
+    assert(roles_state.last_box_status_value ~= nil)
+
+    -- Call on_box_status callbacks after the config is fully applied.
+    call_on_box_status_callbacks(config, roles_state.last_box_status_value)
```

Comment:
If `last_config` is non-nil, then the watcher may call the callbacks. If some of them yield for different (say, random) time intervals, it is possible that `on_box_status` is called before first `apply` for the same role.

I would prefer explicit latch variables: don't call `on_box_status` from the watcher until some `roles_state.post_apply_finished` (and set it after calling the callbacks from here).

Reply:
Also, we discussed with @grafin idea to copy `locked` pattern from `load_cfg.lua` and use it for the `call_on_box_status_callbacks` function: this way we have guaratees similar as for watchers disregarding that we have two source of events: reconfigration and box.status:

> A watcher callback never runs in parallel with itself: if the key to which a watcher is subscribed is updated while the watcher callback is running, the callback will be invoked again with the new value as soon as it returns.

(From the watcher's [doc](https://github.com/tarantool/doc/issues/2407).)

Reply:
Thanks for helping in finding this solution! Implemented.